### ISSUE TITLE
Use a settings file to complement command line argument parsing

### ISSUE THREAD

**Issue by jsvd** [elastic/logstash#4401]
Currently logstash lists the following command line arguments:

```
    -f, --config CONFIG_PATH      Load the logstash config from a specific file
                                  or directory.  If a directory is given, all
                                  files in that directory will be concatenated
                                  in lexicographical order and then parsed as a
                                  single config file. You can also specify
                                  wildcards (globs) and any matched files will
                                  be loaded in the order described above.
    -e CONFIG_STRING              Use the given string as the configuration
                                  data. Same syntax as the config file. If no
                                  input is specified, then the following is
                                  used as the default input:
                                  "input { stdin { type => stdin } }"
                                  and if no output is specified, then the
                                  following is used as the default output:
                                  "output { stdout { codec => rubydebug } }"
                                  If you wish to use both defaults, please use
                                  the empty string for the '-e' flag.
                                   (default: "")
    -w, --filterworkers COUNT     Sets the number of filter workers to run.
                                   (default: 0)
    -l, --log FILE                Write logstash internal logs to the given
                                  file. Without this flag, logstash will emit
                                  logs to standard output.
    -v                            Increase verbosity of logstash internal logs.
                                  Specifying once will show 'informational'
                                  logs. Specifying twice will show 'debug'
                                  logs. This flag is deprecated. You should use
                                  --verbose or --debug instead.
    --quiet                       Quieter logstash logging. This causes only 
                                  errors to be emitted.
    --verbose                     More verbose logging. This causes 'info' 
                                  level logs to be emitted.
    --debug                       Most verbose logging. This causes 'debug'
                                  level logs to be emitted.
    -V, --version                 Emit the version of logstash and its friends,
                                  then exit.
    -p, --pluginpath PATH         A path of where to find plugins. This flag
                                  can be given multiple times to include
                                  multiple paths. Plugins are expected to be
                                  in a specific directory hierarchy:
                                  'PATH/logstash/TYPE/NAME.rb' where TYPE is
                                  'inputs' 'filters', 'outputs' or 'codecs'
                                  and NAME is the name of the plugin.
    -t, --configtest              Check configuration for valid syntax and then exit.
    --[no-]allow-unsafe-shutdown  Force logstash to exit during shutdown even
                                  if there are still inflight events in memory.
                                  By default, logstash will refuse to quit until all
                                  received events have been pushed to the outputs.
                                   (default: false)
    -h, --help                    print help
```

With the introduction of the next generation pipeline and node identification these flags will be added:

```
    -w, --pipeline-workers COUNT  Sets the number of pipeline workers to run.
                                   (default: 2)
    -b, --pipeline-batch-size SIZE Size of batches the pipeline is to work in.
                                   (default: 125)
    -u, --pipeline-batch-delay DELAY_IN_MS When creating pipeline batches, how long to wait while polling
                                  for the next event.
                                   (default: 5)
```

With the pluggable agents, metrics, automatic reloading and more, the big amount of setting flags will clog the cli so I believe it's the time to introduce a settings file.

### Implementation

The settings file should work together with the command line argument parser so that, for example, the user can either specify --node-name in the CLI or "node.name" in the settings file to set the name of the node.

#### Option 1: adding a settings section to the config file

Pros:
  + a single place to check for both the pipeline config and settings
  + consistent syntax between pipeline config and settings

Cons:
  - many people use multiple config files, how to support multiple `config {}` blocks?
  - existing treetop code and AST is painful to manage and change
  - with automatic config reloading and potentially remote loading it's useful to ensure the two concepts are kept separately

#### Option 2: individual YAML file for settings

Pros:
  - easy to understand hierarchical structure to configure core components: pipeline, agent, logger
  - better settings experience across elasticsearch, kibana and beats

Cons:
  - It's YAML

### Common Implementation Problems:

  - No cli arg parsing tool accepts a settings file so the integration between cli args and settings file must be written by us

### Example of YAML file:

```yaml
node:
  name: myindexer
pipeline:
  workers: 2
  batch-size: 125
  batch-delay: 5
  allow-unsafe-shutdown: false
logging:
  verbose: false
  debug: false
  file:
core:
  pluginpath: # $LOGSTASH_HOME/plugins
```

### What now?

What do you think about this? Comments appreciated!

**Comment by jsvd** [comment]
Here's a first stab at implementing a settings file: https://github.com/elastic/logstash/commit/2ed2d4a41df1fb3f9eea7d4bf90b2fbd7bd41bdb

Some concerns:

* having a proper settings file with a yaml hierarchy that can be overridden with the CLI means nearly all CLI arguments will change to make them consistent
* it's not obvious how to cleanly communicate at startup the settings used and the defaults that changed. This is mostly due to the Clamp DSL currently owning the defaults for the settings.

**Comment by suyograo** [comment]
> having a proper settings file with a yaml hierarchy that can be overridden with the CLI means nearly all CLI arguments will change to make them consistent

Discussed with @andrewvc and @jsvd about this - the long option would need to change to match the yaml hierarchy, but we can still keep the existing short notation. Also this change is targeted for the next major version so a breaking change (documented) should be ok

**Comment by jsvd** [comment]
Question:
The location of the settings.yml currently is not configurable and is discoverable as a relative path against LOGSTASH_HOME. In the final implementation, should it:

a) be configurable through command line:
`./bin/logstash -f pipeline_config.txt -s settings.yml`
b) be configurable throught environment variable
SETTINGS_DIR or SETTINGS_FILE
c) be in settings/ or config/ folder?

**Comment by suyograo** [comment]
IMO, we should call this settings file `logstash.yml` and place it under `LS_HOME/conf`. We then introduce a `path.settings` which allows users to override this location either through `logstash.yml` itself or by passing `path.settings` through CLI. This is similar in behavior to Elasticsearch. https://www.elastic.co/guide/en/elasticsearch/reference/current/setup-dir-layout.html

**Comment by suyograo** [comment]
Many users may already have a `LS_HOME/config` directory

**Comment by ph** [comment]
+1 for `$LS_HOME/config/logstash.yml` this is consistent with kibana/es

**Comment by suyograo** [comment]
merged to master and 5.0
